### PR TITLE
Create and clean up a temp dir for use with terraform

### LINKED ISSUE #6873: Using Terraform plugin cache directory option overflows Temp folder in Windows
**_Are you a customer of Octopus Deploy? Don't raise the issue here. Please contact [our support team](https://octopus.com/support) so we can triage your issue, making sure it's handled appropriately._**

# Prerequisites

- [x] I have verified the problem exists in the latest version
- [x] I have searched [open](https://github.com/OctopusDeploy/Issues/issues) and [closed](https://github.com/OctopusDeploy/Issues/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed) issues to make sure it isn't already reported
- [x] I have written a descriptive issue title
- [x] I have linked the original source of this report
- [x] I have tagged the issue appropriately (area/*, kind/bug, tag/regression?)

# The bug

When using `Terraform plugin cache directory` option, one of Windows temporary directories gets larger and larger in size until add disk space is occupied. The issue is present in Terraform CLI itself, relevant issue has been raised [here](https://github.com/hashicorp/terraform/issues/28477).

## What I expected to happen

Temporary files should be deleted after the step has finished.

## Steps to reproduce

1. Set up a runbook or project with a `Apply a Terraform template` step
- All of the default values are fine
- I tested this both by using a package and just using an inline script, both reproduce the issue
- Under Advanced Options, define a path for the Terraform plugin cache directory. e.g. `C:/Users/xxx/AppData/Roaming/terraform-plugin-cache` (Note: The TF docs say you must use forward slashes as path separators even on Windows because their parser requires it - https://www.terraform.io/docs/cli/config/config-file.html#provider-plugin-cache)
2. Run the runbook or project.
3. Check folder `C:\Users\xxx\AppData\Local\Temp` for files with names like `terraform-provider738342411`

## Affected versions

Tested on versions `2021.1` and `2021.2` of Octopus Server.

## Workarounds

`TF_PLUGIN_CACHE_DIR` can now be overridden via user-provided environment variables, which will reduce the required number of provider installations.

No other workaround available other than to periodically clean out the Temp folder.

## Links

[Issue in Terraform CLI](https://github.com/hashicorp/terraform/issues/28477)

### PR DESCRIPTION
**Pull request by mcasperson**
Terraform has an issue where it will not clean up temporary files created while downloading providers: https://github.com/hashicorp/terraform/issues/28477.

By overriding the temporary directory and cleaning it up when Calamari is done, we can work around the the issue.

Here are the provider temporary files in the new directory:
image

Fixes https://github.com/OctopusDeploy/Issues/issues/6873

### CODE REVIEW

File: source/Calamari/Helpers/DisposableDirectory.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+using System;
+using System.IO;
+
+namespace Calamari.Terraform.Helpers
+{
+    public class DisposableDirectory : IDisposable
```
Comment (johnsimons) [inline source/Calamari/Helpers/DisposableDirectory.cs]: you can just use `TemporaryDirectory`

Comment (mcasperson) [reply source/Calamari/Helpers/DisposableDirectory.cs]: Ah, nice. That is a better solution.

**Comment by mcasperson** [comment]
[Terraform doesn't clear out Temp folder after downloading providers](https://trello.com/c/pAWGRDbu/64-terraform-doesnt-clear-out-temp-folder-after-downloading-providers)

**Review by johnsimons — APPROVED** [review]
👍

**Pull request merged**